src/utils/mod.rs

```diff
@@ -63,6 +63,28 @@ pub(crate) fn expand_bits(bit_depth: u8, row_size: u32, buf: &[u8]) -> Vec<u8> {
     p
 }
 
+/// Checks if the provided dimensions would cause an overflow.
+/// 
+/// Returns the optionally provided `ImageError` on overflow or defaults to `LimitErrorKind::DimensionError`.
+#[allow(dead_code)]
+// When no image formats that use it are enabled
+pub(crate) fn check_dimension_overflow(
+    width: u32,
+    height: u32,
+    bytes_per_pixel: u8,
+    error: Option<crate::ImageError>,
+) -> crate::ImageResult<()> {
+    use crate::error::{ImageError, LimitError, LimitErrorKind};
+
+    (width as u64)
+        .checked_mul(height as u64)
+        .and_then(|size| size.checked_mul(bytes_per_pixel as u64))
+        .map(|_| ())
+        .ok_or(error.unwrap_or(ImageError::Limits(LimitError::from_kind(
+            LimitErrorKind::DimensionError,
+        ))))
```

Comment:
A bit too dense for my taste, particularly `map(|_| ())` is weird to read. I've written code like that but now prefer more explicit flow control. So I would split the three cases with `if` after the calculation chain :)

```suggestion
    let byte_len = (width as u64)
        .checked_mul(height as u64)
        .and_then(|size| size.checked_mul(bytes_per_pixel as u64));
        
    if byte_len.is_some() {
        Ok(())
    } else {
        // Or match on `error` with the current parameter type.
        Err(error())
    }
```

Reply:
I took inspiration from the BMP implementation that uses that `map(|_| ())` 😄 . But I agree the chain is a lot. I'll probably go with the [suggested](https://github.com/image-rs/image/pull/1407#discussion_r563167890):
```rust
width as u64 * height as u64 > u64::MAX / bytes_per_pixel as u64
```